settings: Direct loading functionality

rakons opened the pull request:
This commit allows loading data from settings permanent storage
directly to the given callback function.

Note that this commit expands settings API and it is still compatible with the previous version.
New function: settings_load_subtree_direct implemented and proven by unit testing.

The API to the backends is changed and it is not compatible with the old one (additional arguments to the load function). All currently available backends are updated.

rakons committed:
settings: Direct loading functionality

This commit allows loading data from settings permanent storage
directly to the given callback function.

Signed-off-by: Radoslaw Koppel <radoslaw.koppel@nordicsemi.no>

carlescufi commented:
@nvlsianpu can you please take a look at this PR?

nvlsianpu commented on include/settings/settings.h:
```diff
@@ -226,6 +226,59 @@ int settings_load(void);
+ *                        function is called.
+ * @return 0 on success, non-zero on failure.
+ */
+int settings_load_subtree_direct(
```
I would keep both API function. `settings_load_subtree` is function which is consist with initial philosophy injecting settings into modules (push settings).
`settings_load_subtree_direc` introduce first API which pulls settings values from storage.

settings_load_subtree can always be replaced by settings_load_subtree_direc. But disadvantage of this is that it introduce 2 new parameter which are NULL for such a case.

Vudentz replied:
I think it is much simpler to just have it as parameter, it is easier to understand when reading the function documentation then trying to introduce a new one and having to cross reference one another as alternatives.

nvlsianpu replied:
@Vudentz Prefer to keep this as is, as mentioned above I am convinced that it is OK.

nvlsianpu commented on subsys/settings/src/settings_store.c:
```diff
@@ -60,13 +60,34 @@ int settings_load_subtree(const char *subtree)
 	return rc;
 }
 
+int settings_load_subtree_direct(
```
@rako what do you think about making `settings_load_subtree()` the wraper around `settings_load_subtree_direct()`

nvlsianpu replied:
ok, never mind

nvlsianpu commented:
cc @pdunaj

Laczen commented:
I like this functionality.

As it can be used for any variable or subtree that potentially has no registered set routine my proposal would be to name it `settings_load_direct(var, cb)`. 

Limiting the cb function to the h_set type could avoid confusion.

nvlsianpu commented:
@rakons ^^ your opinion

nvlsianpu commented:
cc @lemrey

rakons commented:
> @Laczen 
> Limiting the cb function to the h_set type could avoid confusion.

When it comes to the name - we can use any. When it comes to using h_set type - the problem is that I need context here in my handler.

Vudentz commented on include/settings/settings.h:
```diff
@@ -307,12 +360,17 @@ struct settings_store {
-	 * subtree = NULL loads all values.
+	int (*csi_load)(struct settings_store *cs, const char *subtree,
+			settings_load_direct_cb cb, void *param);
+	/**< Loads values from storage limited to subtree defined by subtree.
```
Right, it is already documenting it after the callback, which is quite weird since we normally document it above the member.

rakons commented on subsys/settings/src/settings_nvs.c:
```diff
@@ -113,19 +115,28 @@ static int settings_nvs_load(struct settings_store *cs, const char *subtree)
+
+			ch->h_set(name_argv, rc2, settings_nvs_read_fn,
+				  (void *) &read_fn_arg);
+		}
```
I am currently working on implementing this functionality.

rakons replied:
Done

rakons committed:
settings: Generic function to call set handler

This commit implements generic function to decide
witch functions to call for selected value name with given
loading parameters.

Signed-off-by: Radoslaw Koppel <radoslaw.koppel@nordicsemi.no>

rakons commented:
Anything more I should change of can we merge it?

nvlsianpu reviewed (approved):
@Laczen Can you revisit?

Laczen commented:
@rakons, I think that with the modified load functionality finding a entry where the set does not return 0 will stop loading all other data (at least for nvs). Could you please check this (line 123 nvs_backend: if (ret) {break;}).

Laczen commented:
@rakons, after doing a rethink about this change, is it not possible to add the cb function and cb_context to the read_fn_arg (settings_nvs_read_fn_arg, ...) ? In that case there would be a lot less changes, it would not change any API, ... when setting up the read_fn_arg cb function can be set to NULL, as well as context.

rakons commented:
@Laczen
> read_fn_arg (settings_nvs_read_fn_arg, ...) ?

Do you mean to add another function to the backend and add it into struct settings_store_itf callbacks?

rakons commented:
@Laczen 
> @rakons, I think that with the modified load functionality finding a entry where the set does not return 0 will stop loading all other data (at least for nvs). Could you please check this (line 123 nvs_backend: if (ret) {break;}).

Yes - it would stop loading if h_set returns non 0 (see settings_call_set_handler in line 193 in settings.c file). Should it behave other way? If it just cannot find a settings handler - such an error is ignored (return 0).

Laczen commented:
> @Laczen
> 
> > @rakons, I think that with the modified load functionality finding a entry where the set does not return 0 will stop loading all other data (at least for nvs). Could you please check this (line 123 nvs_backend: if (ret) {break;}).
> 
> Yes - it would stop loading if h_set returns non 0 (see settings_call_set_handler in line 193 in settings.c file). Should it behave other way? If it just cannot find a settings handler - such an error is ignored (return 0).

@rakons, Yes I think it should behave differently, when a h_set returns non 0 (this could be because the h_set does not find a match to do something or the length is wrong or ...) this should not influence other loads. The error should be ignored.

rakons commented:
@Laczen 
> when a h_set returns non 0 (this could be because the h_set does not find a match to do something or the length is wrong or ...) this should not influence other loads. The error should be ignored.

I see. I will fix it in ``settings_call_set_handler``.

Laczen commented:
> @Laczen
> 
> > read_fn_arg (settings_nvs_read_fn_arg, ...) ?
> 
> Do you mean to add another function to the backend and add it into struct settings_store_itf callbacks?

No, I mean to use the read_fn_arg to pass other data like cb function and cb context. The classical load would keep the cb function and cb context set to NULL, the load direct would change the cb function and cb context to pass the appropriate cb and context.

rakons commented:
I am not sure about it - it seems like just an NVS backend functionality. Isn't it?

Laczen commented:
> I am not sure about it - it seems like just an NVS backend functionality. Isn't it?

Probably I was looking at nvs to much and the proposal was indeed too specific. So your reply was indeed correct, add it to settings_store_itf as csi_load_cb and csi_load_cb_arg, and use this to pass information around.

Vudentz reviewed (approved):
LGTM
